### ISSUE TITLE
Extremely slow pull mcr.microsoft.com/dotnet/sdk:5.0 in Moscow (Russia)

### ISSUE THREAD

**Issue by Gladskih** [microsoft/containerregistry#55]
I'm experiencing extremely slow image pull in Moscow (Russia)
[build 1/7] FROM mcr.microsoft.com/dotnet/sdk:5.0@sha256:48922a379d1f54390d5b8655933d3a62877c1fd46d44ab39cb4133fb1d230681
sha256:e98a60ea39354b242a6cddc260cf41743d5738890323240e609fb91369cc5c19 **17.07MB / 17.07MB 290.2s done**
sha256:62b021f1f14578ffb7f9b9860e9bef156bd26bda5dd16c671d8ce850a5b88110 **155B / 155B 290.8s done**
...
sha256:018bf2f450f77abab99a693b26a25fcbd9f6ea985533a8448817574a1fb3624d **103.28MB / 103.28MB 911.2s done**

**Comment by bariscaglar** [comment]
@Gladskih are you still experiencing the same issues? Sounds like an intermittent CDN issue in your location.

**Comment by Gladskih** [comment]
@bariscaglar Yes. I just try to store base images in cache or do update in background of other daily job. I totally cannot rely on Microsoft registry.

**Comment by bariscaglar** [comment]
One more question: 
It looks like you were pulling these layers at 1Mbit/sec. Can you confirm that pulling a similar image, say from DockerHub, you get better transfer speeds?

We will look into speeds from the CDN provider in your region (Moscow) and compare to the 1Mbit/second you are observing.

**Comment by qweeah** [comment]
@Gladskih It depends on which data endpoint  you are landing on, can you:

1) Provide the domain name in the result of `curl https://mcr.microsoft.com/v2/mcr/hello-world/blobs/sha256:1b930d010525941c1d56ec53b97bd057a67ae1865eebf042686d2a2d18271ced`

2) If the domain name is southeastasia.data.mcr.microsoft.com, provide the result of `nslookup southeastasia.data.mcr.microsoft.com`

**Comment by amityagov** [comment]
Hello everyone!

I can confirm this. Really slow pulling images from mcr.microsoft.com and its lasts for months.

Russia, Yaroslavl

e. g.
> sha256:89483de95a4ca662ae07a0bd29883f07e3fc2db05b489cb705b60e0e969755d7 84.93MB / 105.36MB 364.3s

**Comment by qweeah** [comment]
@amityagov To investigate, could you provide below information on your PC and the VPS? Thanks!

1. Domain name in the non-redirected result of the below command(it should be like *.data.mcr.microsoft.com):
```bash
curl https://mcr.microsoft.com/v2/mcr/hello-world/blobs/sha256:1b930d010525941c1d56ec53b97bd057a67ae1865eebf042686d2a2d18271ced
```

2. DNS lookup result of the domain name:
    E.g. If the domain name is eastasia.data.mcr.microsoft.com, provide the result of `nslookup eastasia.data.mcr.microsoft.com`

**Comment by amityagov** [comment]
@qweeah 

1.
* PC - westeurope.data.mcr.microsoft.com
* VPS - westeurope.data.mcr.microsoft.com

2.
* PC - Server:  dns.google
Address:  8.8.8.8

Non-authoritative answer:
Name:    a-0016.a-msedge.net
Address:  204.79.197.219
Aliases:  eastasia.data.mcr.microsoft.com
          mcr-eastasia.trafficmanager.net
          eastasia-data-mcr-microsoft-com.a-0016.a-msedge.net

* VPS - Server:         127.0.0.53
Address:        127.0.0.53#53

Non-authoritative answer:
eastasia.data.mcr.microsoft.com canonical name = mcr-eastasia.trafficmanager.net.
mcr-eastasia.trafficmanager.net canonical name = eastasia-data-mcr-microsoft-com.a-0016.a-msedge.net.
eastasia-data-mcr-microsoft-com.a-0016.a-msedge.net     canonical name = a-0016.a-msedge.net.
Name:   a-0016.a-msedge.net
Address: 204.79.197.219

**Comment by qweeah** [comment]
You are hitting our Azure Front Door CDN resources in West Europe .

@amityagov Can you ran `traceroute westeurope.data.mcr.microsoft.com` from your PC and VPS and paste the result, Thanks.

**Comment by amityagov** [comment]
@qweeah

VPS

traceroute westeurope.data.mcr.microsoft.com
traceroute to westeurope.data.mcr.microsoft.com (204.79.197.219), 30 hops max, 60 byte packets
 1  * * *
 2  10.82.8.244 (10.82.8.244)  2.163 ms 10.82.8.162 (10.82.8.162)  2.544 ms 10.82.8.160 (10.82.8.160)  2.508 ms
 3  138.197.250.106 (138.197.250.106)  2.536 ms 138.197.250.112 (138.197.250.112)  2.485 ms 138.197.250.114 (138.197.250.114)  2.474 ms
 4  138.197.250.88 (138.197.250.88)  2.447 ms * *
 5  * ams-ix-2.microsoft.com (80.249.209.21)  3.257 ms *
 6  ae22-0.icr01.ams30.ntwk.msn.net (104.44.232.167)  3.301 ms * *
 7  * be-120-0.ibr02.ams30.ntwk.msn.net (104.44.22.223)  3.145 ms *
 8  * * *
 9  ae23-0.ier02.dus30.ntwk.msn.net (104.44.234.173)  8.016 ms be-102-0.ibr01.ams21.ntwk.msn.net (104.44.22.237)  8.777 ms *
10  * * *
11  * * *
12  * * *
13  * * *
14  * * *
15  * * *
16  * * *
17  * * *
18  * * *
19  * * *
20  * * *
21  * * *
22  * * *
23  * * *
24  * * *
25  * * *
26  * * *
27  * * *
28  * * *
29  * * *
30  * * *

PC
 tracert westeurope.data.mcr.microsoft.com

Tracing route to a-0016.a-msedge.net [204.79.197.219]
over a maximum of 30 hops:

  1    <1 ms    <1 ms    <1 ms  192.168.155.1
  2     2 ms     1 ms     2 ms  *-*-*-*.broadband.corbina.ru [128.75.160.1]
  3     *        *        *     Request timed out.
  4     4 ms     1 ms     2 ms  mosk53-bb-po2.yar.corbina.net [85.21.208.57]
  5     5 ms     2 ms     2 ms  79.104.58.233
  6     6 ms     6 ms     6 ms  pe26.Moscow.gldn.net [79.104.225.59]
  7     9 ms     6 ms    12 ms  79.133.94.225
  8     9 ms     8 ms     9 ms  87.226.183.85
  9    54 ms     *        *     46.61.227.174
 10     *       53 ms     *     ae22-0.ear03.fra30.ntwk.msn.net [104.44.235.1]
 11    54 ms    51 ms    51 ms  be-25-0.ibr04.fra30.ntwk.msn.net [104.44.33.139]
 12    50 ms     *        *     ae24-0.ear04.fra30.ntwk.msn.net [104.44.33.142]
 13    48 ms    48 ms    48 ms  ae26-0.fra-96cbe-1b.ntwk.msn.net [104.44.235.25]
 14    53 ms    53 ms    53 ms  13.104.140.67
 15     *        *        *     Request timed out.
 16    51 ms    51 ms    50 ms  204.79.197.219

Trace complete.

**Comment by qweeah** [comment]
@amityagov It's all about network geo-location. Your VPS enjoys a fast download speed because the DigitalOcean Data Center is located in German, directly hop to Microsoft back bone network in Netherlands with latency less that 4ms. But for you PC it tooks 10 hops to reach Microsoft data center in German with 10 times of latency.

MCR's data endpoint is built upon Azure Front Door service, whose presence depends on the location of Azure Data Centers. Since Russia is not in the [officially available regions](https://azurecomcdn.azureedge.net/cvt-953440fb83ee1581804919be2e1842d1be1a551a2b0b11d7c78fdc4c65a1b241/images/shared/regions-map-mobile.svg) in Azure. For now there isn't much we can do to improve the download speed on your PC, I will leave this issue open and let you know if we made any progress.

**Comment by qweeah** [comment]
Hi @amityagov, I was talking to the upstream infra team and they can help take a look at this issue. Can you help run the following command, wait for the downloading(26MB) completes and collect the `x-msedge-ref` header in the response of redirected request, on your PC:

```bash
curl "https://mcr.microsoft.com/v2/windows/servercore/iis/blobs/sha256:b09c5c22d1357e2bb5decff4e1f279067912c337343a5828db551e0f946e0ba6" -vL --output /dev/null
```
Thanks in advance for helping out.

**Comment by amityagov** [comment]
Hi @qweeah

`0000: X-MSEdge-Ref: Ref A: 5EBB5BF061E74F83AC87A753CA7156D5 Ref B: FRA`

**Comment by qweeah** [comment]
> Ref A: 5EBB5BF061E74F83AC87A753CA7156D5

Should be something like  `x-msedge-ref: Ref A: B2F205AD8D10436F96066A2785EB8A99 Ref B: SG2EDGE1518 Ref C: 2021-06-25T06:23:01Z`, can you copy the full header? Thanks

**Comment by amityagov** [comment]
```
0000: X-MSEdge-Ref: Ref A: 5EBB5BF061E74F83AC87A753CA7156D5 Ref B: FRA
0040: EDGE1414 Ref C: 2021-06-25T09:36:52Z
```
